### ISSUE TITLE
[pip] wanted: PTX/JIT build and even better binary wheels

### ISSUE THREAD

**Issue by stas00** [facebookresearch/fairscale#264]
## 🚀 Feature

@sgugger integrated the sharded feature of faiscale into transformers (thank you @blefaudeux and @msbaines for your help!) 
https://github.com/huggingface/transformers/pull/9139
and added a few basic tests: https://github.com/huggingface/transformers/pull/9177
but we won't be able to run those on CIs, since building fairscale via pip can be very slow (as the latest pip does build isolation and fetches pytorch an extra time before doing the build) and there are also build issues.

Would you consider making a JIT build so no binary build is needed - it'd be much easier to install on CI I think - I could be wrong. Since the compilation will still need to happen at run time, we aren't saving time here, but we are saving a ton of time bypassing the recent pip change which made things complicated and error prone (https://github.com/facebookresearch/fairscale/pull/249). 

Or even better to start making binary wheels, but those would still need a PTX/JIT support unless you're going to build those for all cuda compute capabilities.

Bottom line we need a quick and reliable way to get fairscale onto the CI.

If this resonates please have a look at how deepspeed already does PTX builds (it requires no compilation by default on pip install) (though it's a bit buggy at the moment)

Also, this is not urgent at all. I'm just noting that we wrote the tests, but that they won't be running on CI until this is sorted out.

And thank you!

**Comment by msbaines** [comment]
This is a great idea. I will take a look at DeepSpeed. If you have pointers to the specific DeepSpeed code I should look at that would be appreciated. I've not done a PTX build before.

I'll probably start with JIT compiling, as documented here:

https://pytorch.org/tutorials/advanced/cpp_extension.html

**Comment by stas00** [comment]
Thank you!

I've only studied and tweaked parts of their build, so I don't fully know how it is put together, but your entry points would be:

* https://github.com/microsoft/DeepSpeed/blob/master/setup.py
* https://github.com/microsoft/DeepSpeed/blob/master/op_builder/builder.py

and check at how it builds the package if `DS_BUILD_OPS` and its counterparts aren't set - when they are set it builds the binary, otherwise it leaves everything for runtime. So I'd start around here:

https://github.com/microsoft/DeepSpeed/blob/7435b2f10af773b0204e77c3549b2b7df9a7a65b/setup.py#L71

and PTX here:

https://github.com/microsoft/DeepSpeed/blob/7435b2f10af773b0204e77c3549b2b7df9a7a65b/op_builder/builder.py#L217

**Comment by min-xu-ai** [comment]
Is this still an issue?

**Comment by stas00** [comment]
AFAIK, there was no change on your side.

Unless you have things cached the modern pip fetches all  packages it depends on and builds in isolation which take quite some time and it's not guaranteed to fetch the right version of the packages, since it looks at the version dependencies of `fairscale` and completely ignores what gets installed.

So besides all this taking really long time, it builds fairscale against wrong versions of pytorch - it builds it against the version that it fetches and not the version that is already installed. Which precludes the ability to use non-official releases of pytorch.

In my case I have to build it manually:
```
rm -r dist build
python setup.py bdist_wheel
pip uninstall -y fairscale
pip install dist/fairscale-*.whl
```
as `pip install fairscale` completely ignores my environment and breaks it by installing wrong things.

`Deepspeed` doesn't build at install time and instead builds the first time it's being used and only the module that is being needed, so its install takes a few seconds and it doesn't force-fetch any unnecessary dependencies.

As I mentioned the main cause of the problem is that recent switch of `pip` to mimic conda to do build-isolation.

**Comment by stas00** [comment]
It's very fast if I do:
```
pip install fairscale --no-build-isolation -e . -v
```
despite building 2 CUDA extensions.

**Comment by min-xu-ai** [comment]
> as pip install fairscale completely ignores my environment and breaks it by installing wrong things.

I see. Very helpful. Thanks! Mandeep is on PTO. I was scanning all issues. Sounds like we lack some kind of tests around the pip install/build process. I am not yet an expert on pip install/build process. I will read up on it. If you have a small enough test case that can be helpful. Nevermind if you don't I will try figure something out.

**Comment by stas00** [comment]
As I said none of this is your problem but a recent change in how pip builds packages.

I have all the workarounds to make things work. 

I was just proposing that if `pip install fairscale` didn't pre-build the CUDA extensions by default at install time and did that instead at run time via JIT/PTX like Deepspeed does then the install would be much simpler and faster for the user, since it'd already have the right pytorch installed and it just needs to build the extension when it's first needed - so completely transparent to the user and very fast.

Of course, the current pre-build everything option should remain for those who need it.

If I can be of any help please don't hesitate to ask. I'm not sure what test case you're asking about, other than:
```
pip uninstall fairscale -y; pip install fairscale --no-cache --force-reinstall
```
the last two flags are for you to see what happens the first time.

You can compare it to:
```
pip install deepspeed
```
which builds nothing at this stage and builds on the fly at run-time.

**Comment by min-xu-ai** [comment]
I think I understand this better now. Thanks Stas!

I can confirm that `pip install fairscale` on a clear venv take long time. It downloads torch and source tar file of fairscale and build everything. Like Stas said, it likely also build for the wrong torch version.

@blefaudeux, you have worked a bit on the release with this file:

https://github.com/facebookresearch/fairscale/blob/master/.github/workflows/build_wheels.yml

It seems to build the wheel files and upload somewhere.

However, our pypi page only has tar source file:

https://pypi.org/project/fairscale/#files

It is unlike another project like numpy, which has binary wheel files:

https://pypi.org/project/numpy/#files

Perhaps we are building wheel files but uploading to some other places. Once we figure out this mystery, I think we decide how to structure our released artifacts.

**Comment by myleott** [comment]
They don’t auto upload. You have to download the artifacts and upload them when uploading to pypi, see https://github.com/facebookresearch/fairscale/actions/runs/591717171

**Comment by stas00** [comment]
Just a quick word of warning - beside you needing to do a lot more work to build wheels for all the different pytorch versions, pre-built wheels may not work for all or be efficient:
- I don't know if you do any compile time optimization like AVX (efficiency / incompatibility)
- the prebuilt version will have to ensure it includes PTX codes for all possible capabilities.
- There will probably be other binary incompatibilities

So while I initially wrote "... and even better binary wheels" in the name of this Issue, I think it'll be troublesome and will waste your valuable time.

That's why building the first time it's run would be the most efficient approach. While leaving a mechanism for users to pre-build from scratch at install time should they want to do that. Just please make sure that if you do run-time build the binary gets saved into the python env and not under `~/.cache/torch_extensions/` like Deepspeed does it now, since it prevents multi-python environment use. https://github.com/microsoft/DeepSpeed/issues/730

And I'm aware that on certain setups the python env could be read-only, so `/.cache/torch_extensions/fairscale` could be a fallback if that is the case.

**Comment by min-xu-ai** [comment]
Let's say we don't pre-build the cpp part. Just building the wheel file with python code only for py36/37/38/39 and one wheel per py version for all supported torch version. I.e. each release have total 5 files tar + 4 wheel for 4 py versions (well I guess multiple linux/windows/mac too?). Will the 12 extra wheel files be worth it and faster for the users than a single tar file? (again, assuming we do JIT compile for the cpp part.)

**Comment by stas00** [comment]
I'm pretty sure you need only one wheel if you don't build the cpp part - it's pure python + some Cpp files that you will use later - 

So your wheel will be named `fairscale-0.3.0-py3-none-any.whl`

**Comment by min-xu-ai** [comment]
FairScale doesn't build cpp part by default now. I think we can close this. Feel free to reopen if I shouldn't have.
